12272020/root/opt/joy2key/RetroPie-Setup/scriptmodules/supplementary/runcommand/joy2key.py

```diff
@@ -0,0 +1,296 @@
+#!/usr/bin/python3
+# This file is part of The RetroPie Project
+# Modified for ArkOS by ridgek, Dec 2020
+#
+# The RetroPie Project is the legal property of its developers, see
+# https://raw.githubusercontent.com/RetroPie/RetroPie-Setup/master/COPYRIGHT.md
+#
+# The RetroPie Project and this script are released under a GPLv3 license, see
+# https://raw.githubusercontent.com/RetroPie/RetroPie-Setup/master/LICENSE.md
+import os, sys, struct, time, fcntl, termios, signal
+import curses, errno, re
+from pyudev import Context
+
+
+#    struct js_event {
+#        __u32 time;     /* event timestamp in milliseconds */
+#        __s16 value;    /* value */
+#        __u8 type;      /* event type */
+#        __u8 number;    /* axis/button number */
+#    };
+
+JS_MIN = -32768
+JS_MAX = 32768
+JS_REP = 0.20
+
+JS_THRESH = 0.75
+
+JS_EVENT_BUTTON = 0x01
+JS_EVENT_AXIS = 0x02
+JS_EVENT_INIT = 0x80
+
+CONFIG_DIR = '/home/ark/.config/'
+RETROARCH_CFG = CONFIG_DIR + 'retroarch/retroarch.cfg'
+
+def ini_get(key, cfg_file):
+    pattern = r'[ |\t]*' + key + r'[ |\t]*=[ |\t]*'
+    value_m = r'"*([^"\|\r]*)"*'
+    value = ''
+    with open(cfg_file, 'r') as ini_file:
+        for line in ini_file:
+            if re.match(pattern, line):
+                value = re.sub(pattern + value_m + '.*\n', r'\1', line)
+                break
+    return value
+
+def get_btn_num(btn, cfg):
+    num = ini_get('input_' + btn + '_btn', cfg)
+    if num: return num
+    num = ini_get('input_player1_' + btn + '_btn', cfg)
+    if num: return num
+    return ''
+
+def sysdev_get(key, sysdev_path):
+    value = ''
+    for line in open(sysdev_path + key, 'r'):
+        value = line.rstrip('\n')
+        break
+    return value
+
+def get_button_codes(dev_path):
+    js_cfg_dir = CONFIG_DIR + 'all/retroarch-joypads/'
```

Comment:
Changing this line to read the following fixed the mapping for the Odroid Go Advance.  I wonder what happens with this change on the rg351p?  I will test this at a later time.  Good progress so far!

`js_cfg_dir = CONFIG_DIR + 'retroarch/autoconfig/udev/'`

Reply:
OMFG! This directory is empty on my RetroPie except for devices I've manually configured. I didn't even bother to look in the ArkOS verion's! 😂😂😂 It's pre-populated with configs for just about everything ever, including `odroidgo2_joypad` and `OpenSimHardware OSH PB Controller` (the RG351P's device driver). This is EXACTLY what we needed to proceed!

Reply:
note to self, we should symlink this dir like RetroPie does so we don't have to mod the upstream version